main.py

```diff
@@ -1,12 +1,41 @@
-import sys  # used for setting recursion limit higher so program doesn't quit after a while
+import cProfile
+import functools
+import typing
 
 import numpy as np  # used for absolutely everything.
 import time  # used for time reports on minimum solutions and final completion
 import levels_cars as lc  # separate file with every single level setup variable
 from timeit import timeit  # used for line testing
 import itertools  # used for generating different board combinations
 
-sys.setrecursionlimit(100000)
+
+class PrinterProfile(cProfile.Profile):
+    """
+    Profiler context which always prints on context exit
+    if do_print is True
+    """
+    do_print = True
+
+    def __init__(self, do_print=do_print, name=None, print_sort='cumtime'):
+        super().__init__()
+        self.print_sort = print_sort
+        self.name = name
+        self.do_print=do_print
+
+    def __enter__(self):
```

Comment:
These dunder-methods are what makes objects of this class usable as a context manager. when entering a `with` block then `__enter__` is called, and when leaving the block `__exit__` is called, whether it is through normal control low, or an exception